saleor/attribute/migrations/0034_assignedpageattributevalue_page_data_migration.py

```diff
@@ -1,18 +1,52 @@
 # Generated by Django 3.2.20 on 2023-08-24 11:25
 
-from django.db import migrations
-from django.db.models.signals import post_migrate
-from django.apps import apps as registry
+from django.db import migrations, transaction, connection
 
-from .tasks.saleor3_16 import assign_pages_to_attribute_values_task
 
+PAGE_BATCH_SIZE = 5000
 
-def data_migration(apps, _schema_editor):
-    def on_migrations_complete(sender=None, **kwargs):
-        assign_pages_to_attribute_values_task.delay()
 
-    sender = registry.get_app_config("attribute")
-    post_migrate.connect(on_migrations_complete, weak=False, sender=sender)
+def data_migration(apps, _schema_editor):
+    AssignedPageAttributeValue = apps.get_model(
+        "attribute", "AssignedPageAttributeValue"
+    )
+    while (
+        AssignedPageAttributeValue.objects.filter(page__isnull=True)
+        .values_list("pk", flat=True)
+        .exists()
+    ):
+        update_page_assignment()
+
+
+def update_page_assignment():
+    """Update Page assignment.
+
+    Update a batch of 'AssignedPageAttributeValue' rows by setting their 'page' based
+    on their related 'assignment'.
+
+    The number of rows updated in each batch is determined by the BATCH_SIZE.
+    Rows are locked during the update to prevent concurrent modifications.
+    """
+    with transaction.atomic():
+        with connection.cursor() as cursor:
+            cursor.execute(
+                """
+                WITH limited AS (
+                SELECT av.id
+                FROM attribute_assignedpageattributevalue AS av
+                WHERE av.page_id IS NULL
+                ORDER BY av.id DESC
+                LIMIT %s
+                FOR UPDATE
+            )
+            UPDATE attribute_assignedpageattributevalue AS av
+            SET page_id = apa.page_id
+            FROM attribute_assignedpageattribute AS apa
+            WHERE av.id IN (SELECT id FROM limited)
+            AND av.assignment_id = apa.id;
+            """,  # noqa
+                [PAGE_BATCH_SIZE],
+            )
```

Comment:
Instead of making extra SELECT each loop in line 13, cannot we just determine here if there is a need to continue looping? You have a variable here `cursor.rowcount` which can tell you how many rows were affected. When that is zero we can break, WDYT? same would go with `update_product_assignment` below.

Reply:
I've only moved those mutations from the task code, and I don't want to adjust them, unless it's really necessary.

Reply:
Depends on what is the performance gain. If it's a cosmetic change, I think we can keep it as it is right now, but if the performance gain is significant, then we should look into making this improvement here and in other places.

Reply:
For me it sounds more like cosmetic change. Shouldn't affect the performance much